lstests/doc/README.md

```diff
@@ -0,0 +1,688 @@
+
+A `Service` object should know the `Logger` and the settings regarding the launcher service. In this case, except for registering the `Logger`, everything is set to default. Normally it is safe to do so. The `Service` object will automatically detect the file and port if there is an existing launcher service running in the background.
+
+A `Cluster` object should know the `Service` and the settings regarding the cluster of nodes. In this case, the settings include:
+
+1. `node_count=3` : there will be 3 nodes in the cluster
+2. `pnode_count=3` : the 3 nodes all have producers
+3. `producer_count=7` : there will be 7 producers in the cluster
+4. `topology="bridge"` : the network topology will be `bridge`
+5. `center_node_id=1` : node 1 will be the center node for `bridge` topology
+6. `dont_setprod` : do not set producers (using `setprods`) during bootstrapping
+
+The first three settings together suggest that the node mapping will be 3 producer accounts in node #0, and 2 producer accounts each in nodes #1 and #2.
+
+Node #1 is going be to the connecting node for the `bridge` network topology.
+
+In the process of bootstrapping, the producer accounts will be created using `newaccount`, but they will *not* be set as producers using `setprods`. This is because there will be customized producer-setting process in the body of testing.
+
+#### The Context Manager
+
+It is recommended to operate on a `Cluster` object using a context manager, i.e. the `with` statement in the example. With the `with` statement,
+
+```python
+with init_cluster() as clus:
+```
+
+it is guaranteed that when the program crashes, say because of an unexpected `RuntimeError`, all the buffered information will be flushed.
+
+## Configuration
+
+The `Logger`,  `Service` and `Cluster` objects are all configured at their creation.
+
+### Hierarchy
+
+While there are many arguments, the values in general come from three sources, with the latter capable of override the former:
+
+1. default values
+2. in-script arguments, and
+3. command-line arguments.
+
+For example, the default value for the number of nodes in a cluster is 4
+
+```python
+DEFAULT_NODE_COUNT = 4
+```
+
+which can be overridden in the script when creating a cluster
+
+```python
+cluster = Cluster(service=service, node_count=5)
+```
+
+which can be again overridden from the command line. When running the test script, the number of nodes can be changed to 6, by specifying
+
+```bash
+./script.py --node-count=6
+```
+
+or simply
+
+```bash
+./script.py -n 6
+```
+
+The purpose of the design is to allow flexibility in the testing. It allows users to temporarily change the testing behavior without modifying the script. For example, if the user pass `--debug` at the command line, the logger will print out all the logging information at or above `debug` level on the screen, regardless of the original threshold for the log level.
```

Comment:
Copy and paste from my last comment:

> Thanks for the comment. That the number of nodes (and several other similar parameters) is configurable from the command line is primarily to provide convenience to us (the authors). It allows us to quickly debug and test the robustness of `core.service` and the launcher service program behind it. For example, we want to know if we can successfully launch a cluster with 100 nodes at once. Can it work with `mesh` topology? Can it work with `tree` topology? With the command-line options, this can be done with `python -m core.service --node-count 100 --topology tree`. It is also possible that some users may want to try their test cases on different topological types while keeping everything else the same. They may want this flexibility without explicitly writing the config down in the script. Therefore, I strongly recommend that we keep this flexibility.

In general, I recommend that we keep the three-level overriding hierarchy for most of the options, even the reason is to help ourselves for debugging. It did help a lot.